### PR TITLE
[risk=no][RW-6889] Fix bug in test tools

### PR DESCRIPTION
**Pull request by petesantos**
Description:
https://precisionmedicineinitiative.atlassian.net/browse/RW-6889
Fixes an issue in the test tools where using a `setImmediate` call can cause test flakiness when testing UI code that uses a
`setTimeout(fn, 0)`.

According to the[ node documentation](https://nodejs.dev/learn/understanding-setimmediate), it is not clear whether a `setTimeout` of 0 or `setImmediate` will execute first in the next iteration of the event loop.

Running the test in succession demonstrated non-deterministic behavior. After switching to `setTimeout` I no longer saw the issue.

---
**PR checklist**

- [x] This PR meets the Acceptance Criteria in the JIRA story
- [x] The JIRA story has been moved to Dev Review
- [x] This PR includes appropriate unit tests
- [x] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples)
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: ui/src/testing/react-test-helpers.ts [hunk 1]
```diff
@@ -9,7 +9,8 @@ import * as fp from 'lodash/fp';
 // And fails with:
 //   ReferenceError: Zone is not defined
 export async function waitOneTickAndUpdate(wrapper: ReactWrapper) {
-  const waitImmediate = () => new Promise<void>(resolve => setImmediate(resolve));
+  // Make sure to use setTimeout. Combining setTimeout (used in UI) and setImmediate can result in a non-deterministic order of events
```
Comment (calbach) [inline ui/src/testing/react-test-helpers.ts:12]: Which specific callbacks are executing in a non-deterministic order in this test case?

**Comment by calbach** [comment]
I believe I put this context somewhere on a PR - can't find it right now - but IIRC switching to `setTimeout` here causes issues with tests that use fake timers, since `setTimeout` effectively gets monkey-patched by that library while `setImmediate` doesn't. jest/enzyme did not seem to offer a standard way of resolving pending callbacks, from when I last looked, though it's possible you could split out the utilities for these different cases. I believe this will explain many of the errors you see in the UI tests.

File: ui/src/testing/react-test-helpers.ts (hunk 1 above)
Comment (petesantos) [reply ui/src/testing/react-test-helpers.ts:12]: Specifically the [subscribable](https://github.com/all-of-us/workbench/blob/b31d28d9f8dfd531741507952e63edf6027ea4a3/ui/src/app/utils/subscribable.ts#L47) `setTimeout`:

Running this snippet in node repeatedly will demonstrate the issue
```
async function immediatePromise(v) {
  return new Promise(resolve => setImmediate(resolve));
}
async function timeoutPromise() {
  return new Promise(resolve => setTimeout(resolve, 0));
}
function immediate(v) {
  setImmediate(() => console.log('Immediate: ', v))
}
function timeout(v) {
  setTimeout(() => console.log('Timeout: ', v), 0)
}
async function main () {
  console.log('Start Run'.padStart(25, '-').padEnd(50, '-'))
  for (let i = 0; i < 10; i++) {
    // You can play around with different combinations of timeout/immediate to see the different behaviors
    // In general immediate is not deterministic when combined with timeout
    timeout(i)
    immediate(i)
    // await timeoutPromise() 
    // await immediatePromise()
    // console.log('After Immediate: ', i)
  }
}
main();
```

Comment (calbach) [reply ui/src/testing/react-test-helpers.ts:12]: Interesting, thanks for digging into this - analysis looks right to me, though I'm very confused as to how we haven't seen this issue before now.

File: ui/src/testing/react-test-helpers.ts [hunk 2]
```diff
@@ -9,11 +9,18 @@ import * as fp from 'lodash/fp';
...
+  const waitImmediate = () => new Promise<void>(resolve => setImmediate(resolve))
   await act(waitImmediate);
   wrapper.update();
 }
 
+export async function waitOnTimersAndUpdate(wrapper: ReactWrapper){
```
Comment (calbach) [inline ui/src/testing/react-test-helpers.ts:18]: please add a comment to explain when this function should be used vs the other. I think based on what you're saying, the other should only be used with `fakeAsync`, and this should be used in all other cases.

**Review by calbach — APPROVED** [review]

File: ui/src/testing/react-test-helpers.ts (hunk 1 above)
Comment (petesantos) [reply ui/src/testing/react-test-helpers.ts]: I think this is only really an issue when `setTimeout` has a delay of `0`. There aren't too many places in the code where this occurs. Cohort builder seems to do this a bit more than our other components. Most of our timeouts have a delay associated with them.

File: ui/src/testing/react-test-helpers.ts (hunk 2 above)
Comment (petesantos) [reply ui/src/testing/react-test-helpers.ts:20]: added comment

**Pull request merged**